File: stdlib/safe/phrase.go
package safe

import (
	"crypto/hmac"
	"crypto/sha256"
	"io"
	"strings"

	"github.com/art-media-platform/amp.SDK/stdlib/status"
	"golang.org/x/crypto/hkdf"
)

// PhraseChecksumSize is the number of checksum bytes appended before encoding.
// Each word carries 8 bits (PhraseWordCount=256), so a 1-byte checksum adds one word.
const PhraseChecksumSize = 1

// Phrase is an ordered list of canonical wordlist words that encodes a byte
// sequence plus a trailing checksum. Each word carries 8 bits of data.
//
// A Phrase of N words carries (N - PhraseChecksumSize) bytes of entropy.
// The checksum is the leading PhraseChecksumSize bytes of the default
// HashKit digest of the entropy. Decoding fails if the checksum does not match.
type Phrase []string

// EncodePhrase encodes entropy as a Phrase, appending a checksum.
// The returned phrase has len(entropy)+PhraseChecksumSize words.
func EncodePhrase(entropy []byte) Phrase {
	digest := phraseDigest(entropy)
	out := make(Phrase, 0, len(entropy)+PhraseChecksumSize)
	for _, bite := range entropy {
		out = append(out, PhraseWordAt(int(bite)))
	}
	for pos := range PhraseChecksumSize {
		out = append(out, PhraseWordAt(int(digest[pos])))
	}
	return out
}

// DecodePhrase returns the entropy encoded by words after verifying the checksum.
func DecodePhrase(words Phrase) ([]byte, error) {
	if len(words) <= PhraseChecksumSize {
		return nil, status.Code_BadRequest.Errorf("safe: phrase too short (got %d words, need > %d)", len(words), PhraseChecksumSize)
	}
	raw := make([]byte, len(words))
	defer Zero(raw)
	for pos, word := range words {
		idx := PhraseWordIndex(word)
		if idx < 0 {
			return nil, status.Code_BadRequest.Errorf("safe: unknown phrase word %q", word)
		}
		raw[pos] = byte(idx)
	}
	cut := len(raw) - PhraseChecksumSize
	entropy := raw[:cut]
	supplied := raw[cut:]
	expect := phraseDigest(entropy)[:PhraseChecksumSize]
	if !hmac.Equal(supplied, expect) {
		return nil, status.Code_BadRequest.Error("safe: phrase checksum mismatch")
	}
	out := make([]byte, len(entropy))
	copy(out, entropy)
	return out, nil
}

// String returns the phrase as a single space-separated string.
func (phrase Phrase) String() string {
	return strings.Join(phrase, " ")
}

// ParsePhrase splits a whitespace-separated phrase string into a Phrase.
// Case is normalized; surrounding and internal whitespace runs are collapsed.
func ParsePhrase(input string) Phrase {
	return Phrase(strings.Fields(strings.ToLower(input)))
}

// DeriveKey derives a purpose-specific key from the phrase's entropy using
// the default HKDF helper. The phrase is verified as a side-effect; a bad
// checksum returns an error without producing key material.
func (phrase Phrase) DeriveKey(purpose string) ([]byte, error) {
	entropy, err := DecodePhrase(phrase)
	if err != nil {
		return nil, err
	}
	defer Zero(entropy)
	return DeriveSubKey(entropy, purpose)
}

// phraseDigest returns the default HashKit digest of entropy.
func phraseDigest(entropy []byte) []byte {
	kit, _ := NewHashKit(0)
	kit.Hasher.Write(entropy)
	return kit.Hasher.Sum(nil)
}

// KeyPairFromPhrase deterministically derives a KeyPair from phrase entropy.
// Purpose provides domain separation — the same phrase yields distinct keys
// for different roles (e.g. "founder-sig", "device-link", "epoch-seed").
//
// The phrase's checksum is verified before any derivation occurs. The returned
// KeyPair's private material is fresh and owned by the caller; Zero() it after use.
func KeyPairFromPhrase(phrase Phrase, spec KeySpec, purpose string) (KeyPair, error) {
	kit, err := GetKit(spec.CryptoKitID)
	if err != nil {
		return KeyPair{}, err
	}
	entropy, err := DecodePhrase(phrase)
	if err != nil {
		return KeyPair{}, err
	}
	defer Zero(entropy)

	rng := hkdf.New(sha256.New, entropy, nil, []byte(purpose))
	kp := KeyPair{
		Pub: PubKey{
			CryptoKitID: spec.CryptoKitID,
			KeyType:     spec.KeyType,
		},
	}
	if err := generateKeyForSpec(kit, rng, spec.RequestedSize, &kp); err != nil {
		return KeyPair{}, err
	}
	return kp, nil
}

// generateKeyForSpec dispatches GenerateKey to the kit's appropriate capability
// based on KeyType.  Symmetric keys are kit-agnostic (random bytes for both halves).
func generateKeyForSpec(kit *KitSpec, rng io.Reader, requestedSize int, kp *KeyPair) error {
	switch kp.Pub.KeyType {
	case KeyType_SymmetricKey:
		pubSize := requestedSize
		if pubSize < 16 {
			pubSize = 32
		}
		kp.Pub.Bytes = make([]byte, pubSize)
		if _, err := io.ReadFull(rng, kp.Pub.Bytes); err != nil {
			return status.Code_KeyGenerationFailed.Wrap(err)
		}
		kp.Prv = make([]byte, DEKSize)
		if _, err := io.ReadFull(rng, kp.Prv); err != nil {
			return status.Code_KeyGenerationFailed.Wrap(err)
		}
		return nil
	case KeyType_SigningKey:
		if kit.Signing == nil || kit.Signing.Generate == nil {
			return status.Code_Unimplemented.Errorf("KitSpec %s does not generate SigningKeys", kit.ID.String())
		}
		return kit.Signing.Generate(rng, kp)
	case KeyType_AsymmetricKey:
		if kit.Encrypt == nil || kit.Encrypt.Generate == nil {
			return status.Code_Unimplemented.Errorf("KitSpec %s does not generate AsymmetricKeys", kit.ID.String())
		}
		return kit.Encrypt.Generate(rng, kp)
	default:
		return status.Code_Unimplemented.Errorf("unsupported KeyType: %v", kp.Pub.KeyType)
	}
}

File: amp-web/src/crypto/phrase.ts
/**
 * safe.Phrase codec — the canonical identity-seed export vocabulary
 * (SD-did-identity §12.1): a phrase written down in one AMP app restores in
 * any other.
 *
 * Byte-compatible with amp.SDK/stdlib/safe/phrase.go: each word carries 8
 * bits, the trailing PhraseChecksumSize words are the leading bytes of the
 * BLAKE2s-256 digest of the entropy, and purpose-specific keys derive via
 * HKDF-SHA256.  The wordlist is the generated Phrase.Words const
 * (safe.consts.sdl) — never a bespoke list.
 */

import { blake2s } from '@noble/hashes/blake2s';

import { Phrase } from '../generated/safe.consts.js';
import { deriveKey } from './aead.js';

/** Number of words in the canonical wordlist (8 bits/word → byte-aligned entropy). */
export const PhraseWordCount = 256;

/** Checksum bytes appended before encoding; each adds one word to the phrase. */
export const PhraseChecksumSize = 4;

// Wordlist init mirrors phrase.go's phraseLookup guards: exactly
// PhraseWordCount unique words or the module refuses to load.
const wordlist: readonly string[] = Phrase.Words.split(/\s+/).filter((word) => word.length > 0);
const wordIndex = new Map<string, number>();
if (wordlist.length !== PhraseWordCount) {
  throw new Error(`safe: Phrase.Words must contain exactly ${PhraseWordCount} words; got ${wordlist.length}`);
}
for (const [idx, word] of wordlist.entries()) {
  if (wordIndex.has(word)) {
    throw new Error(`safe: Phrase.Words contains duplicate: ${word}`);
  }
  wordIndex.set(word, idx);
}

/** Canonical word at `idx` (0..PhraseWordCount-1).  Throws if out of range. */
export function phraseWordAt(idx: number): string {
  const word = wordlist[idx];
  if (word === undefined) {
    throw new Error(`safe: phrase word index ${idx} out of range`);
  }
  return word;
}

/** Index of `word` in the canonical wordlist, or -1.  Case-sensitive. */
export function phraseWordIndex(word: string): number {
  return wordIndex.get(word) ?? -1;
}

function phraseDigest(entropy: Uint8Array): Uint8Array {
  return blake2s(entropy);
}

/**
 * Encode entropy as a phrase, appending the checksum.  The returned phrase
 * has entropy.length + PhraseChecksumSize words.
 */
export function encodePhrase(entropy: Uint8Array): string[] {
  const digest = phraseDigest(entropy);
  const out: string[] = [];
  for (const bite of entropy) {
    out.push(phraseWordAt(bite));
  }
  for (let pos = 0; pos < PhraseChecksumSize; pos++) {
    out.push(phraseWordAt(digest[pos]));
  }
  return out;
}

/** Decode a phrase back to its entropy after verifying the checksum.  Throws on failure. */
export function decodePhrase(words: readonly string[]): Uint8Array {
  if (words.length <= PhraseChecksumSize) {
    throw new Error(`safe: phrase too short (got ${words.length} words, need > ${PhraseChecksumSize})`);
  }
  const raw = new Uint8Array(words.length);
  try {
    for (const [pos, word] of words.entries()) {
      const idx = phraseWordIndex(word);
      if (idx < 0) {
        throw new Error(`safe: unknown phrase word ${JSON.stringify(word)}`);
      }
      raw[pos] = idx;
    }
    const cut = raw.length - PhraseChecksumSize;
    const entropy = raw.subarray(0, cut);
    const supplied = raw.subarray(cut);
    const expect = phraseDigest(entropy);
    let diff = 0;
    for (let pos = 0; pos < PhraseChecksumSize; pos++) {
      diff |= supplied[pos] ^ expect[pos];
    }
    if (diff !== 0) {
      throw new Error('safe: phrase checksum mismatch');
    }
    return new Uint8Array(entropy);
  } finally {
    raw.fill(0);
  }
}

/**
 * Split a whitespace-separated phrase string into words.  Case is normalized;
 * surrounding and internal whitespace runs are collapsed.
 */
export function parsePhrase(input: string): string[] {
  return input.toLowerCase().split(/\s+/).filter((word) => word.length > 0);
}

/** Join a phrase into a single space-separated string. */
export function phraseToString(words: readonly string[]): string {
  return words.join(' ');
}

/**
 * Derive a purpose-specific 32-byte key from the phrase's entropy
 * (HKDF-SHA256, no salt, purpose as info — safe.DeriveSubKey).  The phrase is
 * verified as a side-effect; a bad checksum throws without producing key
 * material.
 */
export function derivePhraseKey(words: readonly string[], purpose: string): Uint8Array {
  const entropy = decodePhrase(words);
  try {
    return deriveKey(entropy, null, new TextEncoder().encode(purpose));
  } finally {
    entropy.fill(0);
  }
}
